Nlp text.rebased

zurabksmlp opened the pull request:
This pull request is for branch nlp_text.reabsed, it replaces earlier pull request 15 opened on branch nlp_text, called NLP capabillities.

Commit:
spec file for regression test

Commit:
nlp initial commit beforre pull request

Commit:
adding master and data files to nlp branch

Commit:
Added LLM training from scratch, finetuning, RAG, and some agentic capabilites on top of SMLP -- such as ability to accept text as a command and execute it, and also an MCP style invocation of SMLP by an MCP client. These changes are well documented in an updated SMLP manual.

Here are more details:
•	Changes related to moving to tensorflow 2.19.0  and keras 3.8.0, Keras API changed significantly since version keras=3.0.0. Modules affected: train_keras.py, smlp_models.py, and smlp_terms.py.
•	Added initial LLM capabilities: training from scratch, finetuning, and RAG. NLP and text analysis capabilities (text classification, regression, and root cause analysis using subgroup discovery were introduced in previous pool request). Modules: smlp_llm.py, smlp_finetune.py, smlp_rag.py, smlp_generate.py. Also affected: smlp_config.py
•	Added initial agentic capabilities: An SMLP Agent capable to receive command as text and launch SMLP, an SMLP bot and a flow where SMLP Agent interact with another agent. These capabilities are experimental and work in progress. Modules introduced: smlp_agent.py, api_smlp_agent.py, smlp_mcp.py, smlp_chatbot.py.
•	Added initial MCP capabilities to SMLP: created modules smlp_mcp_client.py and smlp_mcp_server.py. The latter is supposed to execute SMLP invocation request sent to it using MCP clients – and smlp_mcp_client.py is such a client that sends request to SMLP_MCP_server.py (currently client-server communication is only supported with local IO – class StdioTransport).
•	Other modules affected: smlp_config.py, smlp_flows.py, smlp_logs.py, and smlp_plots.py (change here was required for MCP support).
•	Regression scripts smlp_regr.py and smp_regr.csv have been updated too: new modes supported, and new tests added.

Virtual env library versions:
Package                     Version
--------------------------- -----------
absl-py                     2.1.0
accelerate                  1.6.0
aiohappyeyeballs            2.4.6
aiohttp                     3.11.13
aiosignal                   1.3.2
altair                      5.5.0
annotated-types             0.7.0
anyio                       4.8.0
asttokens                   3.0.0
astunparse                  1.6.3
attrs                       25.1.0
Authlib                     1.6.1
blinker                     1.9.0
blis                        1.2.0
blobfile                    3.0.0
cachetools                  6.1.0
catalogue                   2.0.10
category_encoders           2.7.0
certifi                     2025.1.31
cffi                        1.17.1
charset-normalizer          3.4.1
click                       8.1.8
cloudpathlib                0.21.0
cloudpickle                 3.1.1
comm                        0.2.2
confection                  0.1.5
contourpy                   1.3.1
cryptography                45.0.6
cycler                      0.12.1
cyclopts                    3.22.5
cymem                       2.0.11
Cython                      3.0.12
dash                        2.18.2
dash-core-components        2.0.0
dash-html-components        2.0.0
dash-table                  5.0.0
dataclasses-json            0.6.7
datasets                    3.5.1
debugpy                     1.8.12
decorator                   5.1.1
deprecation                 2.1.0
dill                        0.3.8
distro                      1.9.0
diversipy                   0.9
dnspython                   2.7.0
docstring_parser            0.17.0
docutils                    0.22
doepy                       0.0.1
eland                       9.0.1
elastic-transport           8.17.1
elasticsearch               8.13.0
email_validator             2.2.0
en_core_web_lg              3.8.0
en_core_web_sm              3.8.0
exceptiongroup              1.3.0
executing                   2.2.0
faiss-cpu                   1.10.0
fastapi                     0.115.13
fastjsonschema              2.21.1
fastmcp                     2.11.3
fasttext                    0.9.3
filelock                    3.18.0
Flask                       3.0.3
flatbuffers                 25.2.10
fonttools                   4.56.0
frozenlist                  1.5.0
fsspec                      2025.3.0
gast                        0.6.0
gensim                      4.3.3
gitdb                       4.0.12
GitPython                   3.1.44
google-pasta                0.2.0
greenlet                    3.1.1
grpcio                      1.70.0
h11                         0.14.0
h5py                        3.13.0
httpcore                    1.0.7
httpx                       0.28.1
httpx-sse                   0.4.0
huggingface-hub             0.30.2
idna                        3.10
imbalanced-learn            0.13.0
importlib_metadata          8.6.1
intel_extension_for_pytorch 2.7.0
ipykernel                   6.29.5
ipython                     8.32.0
ipywidgets                  8.1.5
isodate                     0.7.2
itsdangerous                2.2.0
jedi                        0.19.2
jenkspy                     0.4.1
Jinja2                      3.1.5
jiter                       0.9.0
joblib                      1.3.2
jsonpatch                   1.33
jsonpointer                 3.0.0
jsonschema                  4.23.0
jsonschema-path             0.3.4
jsonschema-specifications   2024.10.1
jupyter_client              8.6.3
jupyter_core                5.7.2
jupyterlab_widgets          3.0.13
kaleido                     0.2.1
keras                       3.8.0
keras-tuner                 1.4.7
kiwisolver                  1.4.8
kt-legacy                   1.0.5
langchain                   0.3.27
langchain-community         0.3.18
langchain-core              0.3.72
langchain-ollama            0.2.3
langchain-text-splitters    0.3.9
langcodes                   3.5.0
langgraph                   0.5.4
langgraph-checkpoint        2.1.1
langgraph-prebuilt          0.5.2
langgraph-sdk               0.1.74
langsmith                   0.3.45
language_data               1.3.0
lazy-object-proxy           1.11.0
libclang                    18.1.1
lightgbm                    4.6.0
llvmlite                    0.44.0
lxml                        5.4.0
marisa-trie                 1.2.1
Markdown                    3.7
markdown-it-py              3.0.0
MarkupSafe                  3.0.2
marshmallow                 3.26.1
matplotlib                  3.7.5
matplotlib-inline           0.1.7
mcp                         1.12.4
mdurl                       0.1.2
meson                       1.7.0
ml_dtypes                   0.5.1
mljar-scikit-plot           0.3.12
more-itertools              10.7.0
mpmath                      1.3.0
mrmr-selection              0.2.8
multidict                   6.1.0
multiprocess                0.70.16
murmurhash                  1.0.12
mypy-extensions             1.0.0
namex                       0.0.8
narwhals                    1.44.0
nbformat                    5.10.4
nest-asyncio                1.6.0
networkx                    3.4.2
nltk                        3.9.1
numba                       0.61.0
numpy                       1.26.4
nvidia-cublas-cu12          12.6.4.1
nvidia-cuda-cupti-cu12      12.6.80
nvidia-cuda-nvrtc-cu12      12.6.77
nvidia-cuda-runtime-cu12    12.6.77
nvidia-cudnn-cu12           9.5.1.17
nvidia-cufft-cu12           11.3.0.4
nvidia-cufile-cu12          1.11.1.6
nvidia-curand-cu12          10.3.7.77
nvidia-cusolver-cu12        11.7.1.2
nvidia-cusparse-cu12        12.5.4.2
nvidia-cusparselt-cu12      0.6.3
nvidia-nccl-cu12            2.26.2
nvidia-nvjitlink-cu12       12.6.85
nvidia-nvtx-cu12            12.6.77
ollama                      0.4.7
openai                      1.78.0
openapi-core                0.19.5
openapi-pydantic            0.5.1
openapi-schema-validator    0.6.3
openapi-spec-validator      0.7.2
opt_einsum                  3.4.0
optree                      0.14.0
orjson                      3.10.15
ormsgpack                   1.10.0
outcome                     1.3.0.post0
packaging                   24.2
pandas                      2.2.3
parse                       1.20.2
parso                       0.8.4
pathable                    0.4.4
patsy                       1.0.1
peft                        0.15.2
pexpect                     4.9.0
pillow                      11.1.0
pip                         25.2
platformdirs                4.3.6
plotly                      5.24.1
plotly-resampler            0.10.0
pmdarima                    2.0.4
polars                      1.22.0
preshed                     3.0.9
prompt_toolkit              3.0.50
propcache                   0.3.0
protobuf                    5.29.3
psutil                      7.0.0
ptyprocess                  0.7.0
pure_eval                   0.2.3
pyarrow                     20.0.0
pybind11                    2.13.6
pycaret                     3.4.0
pycparser                   2.22
pycryptodomex               3.22.0
pydantic                    2.11.7
pydantic_core               2.33.2
pydantic-settings           2.8.1
pydeck                      0.9.1
pyDOE                       0.3.8
Pygments                    2.19.1
PyMuPDF                     1.25.3
pymupdf4llm                 0.0.17
pyod                        2.0.3
pyparsing                   3.2.1
pyperclip                   1.9.0
pysubgroup                  0.8.0
python-dateutil             2.9.0.post0
python-dotenv               1.1.1
python-multipart            0.0.20
pytz                        2025.1
PyYAML                      6.0.2
pyzmq                       26.2.1
referencing                 0.36.2
regex                       2024.11.6
requests                    2.32.3
requests-toolbelt           1.0.0
retrying                    1.3.4
rfc3339-validator           0.1.4
rich                        13.9.4
rich-rst                    1.3.1
rpds-py                     0.22.3
safetensors                 0.5.3
schemdraw                   0.15
scikit-base                 0.7.8
scikit-learn                1.4.2
scikit-plot                 0.3.7
scipy                       1.11.4
seaborn                     0.13.2
sentence-transformers       4.1.0
setuptools                  80.9.0
shellingham                 1.5.4
six                         1.17.0
sklearn-compat              0.1.3
sktime                      0.36.0
smart-open                  7.1.0
smmap                       5.0.2
sniffio                     1.3.1
sortedcontainers            2.4.0
spacy                       3.8.4
spacy-legacy                3.0.12
spacy-loggers               1.0.5
spacy-lookups-data          1.0.5
SQLAlchemy                  2.0.38
srsly                       2.5.1
sse-starlette               3.0.2
stack-data                  0.6.3
starlette                   0.46.2
statsmodels                 0.14.4
streamlit                   1.46.0
sympy                       1.14.0
tbats                       1.1.3
tenacity                    9.0.0
tensorboard                 2.19.0
tensorboard-data-server     0.7.2
tensorflow                  2.19.0
termcolor                   2.5.0
tf_keras                    2.19.0
thinc                       8.3.4
threadpoolctl               3.5.0
tiktoken                    0.9.0
tokenizers                  0.21.1
toml                        0.10.2
torch                       2.7.0
tornado                     6.4.2
tqdm                        4.67.1
traitlets                   5.14.3
transformers                4.52.4
trio                        0.24.0
triton                      3.3.0
trl                         0.17.0
tsdownsample                0.1.4.1
typer                       0.15.2
typing_extensions           4.12.2
typing-inspect              0.9.0
typing-inspection           0.4.1
tzdata                      2025.1
urllib3                     2.3.0
uvicorn                     0.34.3
wasabi                      1.1.3
watchdog                    6.0.0
wcwidth                     0.2.13
weasel                      0.4.1
Werkzeug                    3.0.6
wheel                       0.45.1
widgetsnbextension          4.0.13
wrapt                       1.17.2
wurlitzer                   3.1.1
xxhash                      3.5.0
yarl                        1.18.3
yellowbrick                 1.5
zipp                        3.21.0
zstandard                   0.23.0

fbrausse commented on src/smlp_py/smlp_models.py:
```diff
@@ -292,7 +308,7 @@ def _report_prediction_results(self, algo:str, resp_names:list[str], resp_df:pd.
                                 str(self.predictions_summary_filename(data_version)))
         predictions_df.to_csv(self.predictions_summary_filename(data_version), index=True)
-        #print('predictions_df\n', predictions_df) 
+        #print('predictions_df\n', predictions_df)
```
Do we need this commented code?

fbrausse commented on src/smlp_py/smlp_models.py:
```diff
@@ -192,22 +192,38 @@ def _pred_results_to_df(self, algo, resp_names, resp, pred):
+                        raise Exception('Implementation error in function pred_results_to_df')   
+                pred_ind = resp.index if not resp is None else range(len(pred)); #print('pred_ind', pred_ind)
+                predictions_df = pd.DataFrame(pred, index=pred_ind, columns=predictions_colnames)
+        #print('predictions_df\n', predictions_df)
```
Do we need this commented print statement?

fbrausse commented on src/smlp_py/smlp_plots.py:
```diff
@@ -2,7 +2,12 @@
+if not os.environ.get("DISPLAY"):
+    matplotlib.use("Agg")  # Non-interactive backend
+else:
+    matplotlib.use("TkAgg") # Set the backend to 'TkAgg' before importing pyplot, to enable interactive plots
```
I wish we wouldn't default to this kind of interactive use of SMLP based on some unrelated environment settings. I understand it can be useful, but should, in my opinion, at least be possible to deactivate without messing with envvars.

fbrausse commented on src/smlp_py/smlp_subgroups.py:
```diff
@@ -813,7 +813,6 @@ def single_range_rule_to_row(r):
     # function to visualize the selected ranges).
     def _smlp_subgroups_single_response(self, feat_df:pd.DataFrame, resp_df:pd.DataFrame, 
             resp_name:str, pos_value:int, qf:str, dim:int, top_n:int):
-        assert pos_value == 0 or pos_value == 1
```
Why remove this assertion?

fbrausse commented on src/smlp_py/smlp_utils.py:
```diff
@@ -232,17 +237,29 @@ def list_remove_duplicates(lst):
     except:
         res2 = None
     if res2 is not None:
+        #print(res2, res, type(column)); print(type(column[0])); 
```
Do we need this commented print statement?

zurabksmlp replied:
see issue #26

fbrausse commented on src/smlp_py/smlp_utils.py:
```diff
@@ -232,17 +237,29 @@ def list_remove_duplicates(lst):
+    #assert column.dtype in [int , float , object , datetime , bool]
+    #print('column.dtype', column.dtype)
+    #res = column.dtype in [int, float]
+    #assert res == (column.dtype.name in ['int64', 'float64'])
```
Do we need this commented code?

fbrausse commented on src/smlp_py/train_sklearn.py:
```diff
@@ -310,6 +310,7 @@ def get_sklearn_hparam_default_dict(self):
     def dt_regr_train(self, feature_names, resp_names, algo, hparam_dict,
             X_train, X_test, y_train, y_test, seed, weights):
         # Fit the regressor, set max_depth = 3
+        #print(X_train, '\n', X_test, '\n', y_train, '\n', y_test)
```
Do we need this commented code?

zurabksmlp replied:
See is #26

fbrausse reviewed (commented):
All in all, seems good. Additional comments: There are plenty of new tests comparing the keras version to 3.0.0 using `<=`, i.e., with 3.0.0 included. This looks strange to me: why would old API be present in v3.0.0 and below, but not in 3.0.1 or later?

fbrausse commented on regr_smlp/code/smlp_text.py:
According to email conversion this file should be removed from the PR.

zurabksmlp replied:
issue #27 has been created

fbrausse commented on regr_smlp/code/transformer.py:
Same, this file should be removed from the PR, right?

zurabksmlp replied:
See issue #27

zurabksmlp committed:
Update smlp_subgroups.py

zurabksmlp committed:
Update smlp_utils.py

Commit:
Added support for LLM generated result quality validation using LLM-as-Jusge method. LLM-as-Jusge is applied to results of LLM scratch training, finetuning and RAG flows

Commit:
code changes to support LLM-as-Judge capability

Commit:
added test for LLM-as-Judge flow

Commit:
Merge branch 'nlp_text.rebased' of github.com:fbrausse/smlp into nlp_text.rebased

Commit:
small fixes to SMLP MCP

Commit:
updated gitignore

Commit:
RL enhancemnt to SMLP Agent

Commit:
Merge branch 'master' into nlp_text.rebased

Commit:
Add SMLP RL Agent core files

- smlp_agent_rl_server.py: FastAPI backend with all endpoints
- smlp_agent_utils.py: Shared utilities (LLM calls, execution, filtering)
- smlp_few_shot_examples.json: Initial example pool with UCB scores

Commit:
Complete SMLP RL Agent implementation

- smlp_agent_rl.py: Core RL components (ExampleStore, RewardModel, PromptOptimizer, RLTrainer)
- smlp_agent_web_ui.html: Web interface with JSON editor, star rating, dual checkboxes
- smlp_few_shot_examples.json: Seed example pool with UCB scores
- smlp_py/smlp_config.py: Configuration updates
- .gitignore: Exclude RL runtime files

Commit:
small fixes in SMLP RL Agent

Commit:
Remove obsolete RL Agent documentation and integration files

- INTEGRATION_GUIDE.md: replaced by LaTeX manual
- PROJECT_SUMMARY.md: obsolete development notes
- SMLP_RL_README.md: replaced by LaTeX manual Section (RL Agent)
- smlp_agent_rl_integration.py: obsolete experimental code (agents kept separate)

Commit:
Add comprehensive test suite for SMLP RL Agent

- test_rl_agent.py: 16 unit tests for core RL components (deterministic)
- test_rl_server.py: 16 integration tests for API endpoints (mocked LLM)
- test_e2e_manual.py: 6 end-to-end smoke tests (manual validation)
- README_pytest.md: Testing documentation and usage guide

All tests pass (32/32) with zero warnings. Tests use proper mocking
patterns (patch where used, not where defined) and follow pytest best
practices.

Commit:
Merge master into nlp_text.rebased after resolving conflicts

Commit:
Restore stashed work and resolve merge conflicts

Commit:
Merge master into nlp_text.rebased and resolve conflicts

fbrausse committed:
re-add "extended" manual

Commit:
merge from master

Commit:
removed redundant assertion from smlp_subgroups.py

Commit:
Record regression model JSONs

Commit:
Ignore local model artifacts and update Test228 regression masters

Commit:
update master files

Commit:
update master json files

Commit:
initial support for mlflow based model saving and loading in RAG mode
